Remove MD5 CRL from SSL tests

minfrin opened the pull request:
- Add explicit SERF_SSL_SIGNATURE_FAILURE error code, so that serf does not return unknown error when encountered obsolete algorithms.

- Update create_certs.py to work with python3.

- Update create_certs.py to use sha256 instead of md5 for creation of CRLS.

- Regenerate certificates, keys, and CRLs for testing.

- Update test_ssl.c to expect regenerated certificate fingerprints.

minfrin committed:
Remove MD5 CRL from SSL tests

- Add explicit SERF_SSL_SIGNATURE_FAILURE error code, so that serf
  does not return unknown error when encountered obsolete algorithms.

- Update create_certs.py to work with python3.

- Update create_certs.py to use sha256 instead of md5 for creation
  of CRLS.

- Regenerate certificates, keys, and CRLs for testing.

- Update test_ssl.c to expect regenerated certificate fingerprints.

d-sahlberg commented on buckets/ssl_buckets.c:
```diff
@@ -890,7 +890,11 @@ validate_server_certificate(int cert_valid, X509_STORE_CTX *store_ctx)
+            case X509_V_ERR_CRL_SIGNATURE_FAILURE:
+		    failures |= SERF_SSL_SIGNATURE_FAILURE;
+		    break;
+	    default:
```
Minor nit: indentation issues?

d-sahlberg commented on test/certs/create_certs.py:
```diff
@@ -73,18 +73,18 @@ def create_crl(revokedcert, cakey, cacert, crlfile, next_crl_days=VALID_DAYS):
     try:
-        exported = crl.export(cacert, cakey, days=next_crl_days, digest=b"md5")
+        exported = crl.export(cacert, cakey, days=next_crl_days, digest=b"sha256")
     except TypeError:
```
I believe most changes here are related to Py3 except this one. Can we split it up to separate commits, just for clarification?

brainy replied:
RIght. There are two completely unrelated changes here: The MD5 -> SHA256 change + regenerate all certificates; and the Py3-compat change. These should be split in two commits.

Regarding the change to create_certs.py, I also recommend  setting the notBefore and notAfter dates explicitly to "Oct  4 22:44:52 2021 GMT" and "Sep 10 22:44:52 2121 GMT" and regenerate the certs (whiles still using MD5). This is a sanity check: the certificates should not change at all, and the related changes in test_ssl.c should not be necessary. The cert fingerprint should only change when you update to SHA-256.

brainy replied:
(By the way, that script still uses pyOpenSSL ... highly recommended to use python-cryptography instead. But that's not the topic here.)

minfrin replied:
One of the difficulties is testing this, which is why all the changes are related.

With broken CRLs the tests fail. Recreating the certs to make the tests pass breaks because create_certs.py won't compile/run.

I can break these up, but don't want to have to explain why the tests fail until the patches are brought together, especially when people are looking for extra review.

d-sahlberg reviewed (changes requested):
In general I think this looks good although I'm not an expert and would like additional review from the community.

I'd prefer if we can split it up to separate parts, one for Py3, one for changing the CRL-algorithm and one for the error code change.

minfrin committed:
Begone foul tabs

And a plague upon the house of vi for adding tabs to a file already
exclusively indented with spaces.

d-sahlberg reviewed (commented):
> And a plague upon the house of vi for adding tabs to a file already exclusively indented with spaces.
Don't know if you run oldschool hardcore vi but doesn't at least vim support https://editorconfig.org/?

d-sahlberg commented:
I've committed the py3 changes in r1926623.
However, as Brane has already written, it would be good to update it to python-cryptography. In particular I had problems with crypto.PKCS12, which was removed from pyopenssl in 24.1, I had to explicitly install pyopenssl==24.0 to test these changes.

d-sahlberg commented:
@brainy Are you happy with the new SERF_SSL_SIGNATURE_FAILURE error code?

I intend to commit this in two more batches:
- SERF_SSL_SIGNATURE_FAILURE, basically buckets/ssl_buckets.c and serf_bucket_types.h plus the second part of test/test_ssl.c
- The regenerated certificates (including the argument to create such certs in test/certs/create_certs.py and the new fingerprints in test/test_ssl.c).

Does this sound reasonable to everyone?

chemodax commented on buckets/ssl_buckets.c:
```diff
@@ -890,6 +890,10 @@ validate_server_certificate(int cert_valid, X509_STORE_CTX *store_ctx)
             case X509_V_ERR_UNABLE_TO_GET_CRL:
                     failures |= SERF_SSL_CERT_UNABLE_TO_GET_CRL;
                     break;
+            case X509_V_ERR_CERT_SIGNATURE_FAILURE:
+            case X509_V_ERR_CRL_SIGNATURE_FAILURE:
+                    failures |= SERF_SSL_SIGNATURE_FAILURE;
+                    break;
```
This looks like unrelated change.

Also I suggest to add separate flag for Certificate and CRL signature failure.

d-sahlberg replied:
Unrelated to the switch from md5 to sha256, yes. That's why I'd commit separately (together with the change in test_ssl.c).

minfrin replied:
On the separate flag, openssl returns 95 possible reasons for a certificate not to verify, and we ignore most of them, returning "unknown error", which translates to "an error has occurred".

https://github.com/openssl/openssl/blob/cdd01b5e0734b0324251b32a8edd97f42ba90429/include/openssl/x509_vfy.h.in#L206

Ideally we want to expose underlying errors as strings, accessible to caller above.

It should not be necessary to know or care what crypto API created the string, that string must be visible to the caller.

What I had in mind was to attach the last known SSL error to the pool, or to a context, and have a function like this:

```
const char *serf_error_detail(apr_pool_t pool);
```

chemodax commented on serf_bucket_types.h:
```diff
@@ -583,6 +583,8 @@ serf_bucket_t *serf_bucket_limit_create(
 #define SERF_SSL_OCSP_RESPONDER_ERROR           0x0200
 #define SERF_SSL_OCSP_RESPONDER_UNKNOWN_FAILURE 0x0400
 
+#define SERF_SSL_SIGNATURE_FAILURE  0x0800
+
```
Indentation -- values should be aligned.
